elf/arch-loongarch.cc

```diff
@@ -364,11 +364,73 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
       write_k12(loc, highest12(S + A, P));
       break;
     case R_LARCH_GOT_PC_LO12:
-      write_k12(loc, GOT + G + A);
+      if (i >= 2 && get_r_delta(i-1) - get_r_delta(i-2) == 4) {
+        // pcalau12i/ld.d has been relaxed to pcalau12i/addi.d, and
+        // then the pair has been relaxed to pcaddi.
+        // loc stores 'ld.d', rewrite ld.d with pcaddi
+        *(ul32 *)loc = 0x1800'0000 | get_rd(*(ul32 *)loc);
+        write_j20(loc, (S + A - P) >> 2);
+      } else {
+        if (i >= 2 &&
+          i + 1 < rels.size() &&
+          sym.is_local(ctx) &&
+          !sym.is_absolute() &&
+          !sym.is_ifunc() &&
+          ctx.arg.relax &&
+          rels[i-1].r_type == R_LARCH_RELAX &&
+          rels[i+1].r_type == R_LARCH_RELAX &&
+          rels[i-2].r_type == R_LARCH_GOT_PC_HI20 &&
+          rels[i-2].r_offset == rel.r_offset - 4) {
+          u32 insn1 = *(ul32 *)(contents.data() + rels[i-2].r_offset);
+          u32 insn2 = *(ul32 *)(contents.data() + rel.r_offset);
+          u32 rd = get_rd(insn1);
+
+          if (rd == get_rd(insn2)) {
+            // pcalau12i/ld.d has been relaxed to pcalau12i/addi.d
+            // rewrite the ld.d with addi.d
+            *(ul32 *)loc = 0x02c00000 | rd | (rd << 5);
+            write_k12(loc, S + A);
+            break;
+          }
+        }
+
+        // relax not applied.
+        write_k12(loc, GOT + G + A);
+      }
       break;
     case R_LARCH_GOT_PC_HI20:
-      write_j20(loc, hi20(GOT + G + A, P));
-      break;
+      switch (removed_bytes) {
+        // pcalau12i/ld.d has been relaxed to pcaddi, the first insn has been removed.
+        case 4:
+          break;
+        case 0:
+          if (i + 3 < rels.size() &&
+            sym.is_local(ctx) &&
+            !sym.is_absolute() &&
+            !sym.is_ifunc() &&
+            ctx.arg.relax &&
+            rels[i+1].r_type == R_LARCH_RELAX &&
```

Comment:
You want to pay more attention to the details -- add spaces around `+`. This is how we format it in other places.